File: go/pkg/hey/clearances.go
package hey

import (
	"context"
	"fmt"

	"github.com/basecamp/hey-sdk/go/pkg/generated"
)

// ClearancesService works the Screener: who is waiting to be let in, and letting them in
// or turning them away.
type ClearancesService struct {
	client *Client
}

// NewClearancesService creates a new ClearancesService.
func NewClearancesService(client *Client) *ClearancesService {
	return &ClearancesService{client: client}
}

// PendingCount answers how many senders are waiting, without fetching them.
//
// This is the cheap read HEY's own apps sync for the Screener badge. Use Pending when you
// want the senders themselves, or Summary when you also want the stream to follow.
func (s *ClearancesService) PendingCount(ctx context.Context) (count int, err error) {
	summary, err := s.Summary(ctx)
	if err != nil {
		return 0, err
	}
	if summary == nil {
		return 0, nil
	}
	return int(summary.PendingClearancesCount), nil
}

// Summary answers everything HEY says about the Screener without the queue itself: how
// many senders are waiting, and the signed stream name to subscribe to on HEY's cable
// server to be told when that changes.
//
// It is the same read as PendingCount — the count alone, no queue dragged along.
func (s *ClearancesService) Summary(ctx context.Context) (summary *generated.ClearanceSummary, err error) {
	op := OperationInfo{
		Service: "Clearances", Operation: "GetClearances",
		ResourceType: "clearance", IsMutation: false,
	}

	err = s.client.instrument(ctx, op, func(ctx context.Context) error {
		resp, rerr := s.client.genClient().GetClearancesWithResponse(ctx, &generated.GetClearancesParams{})
		if rerr != nil {
			return rerr
		}
		if cerr := CheckResponse(resp.HTTPResponse); cerr != nil {
			return cerr
		}
		summary = resp.JSON200
		return nil
	})
	return summary, err
}

// ClearancePage contains one page of clearances and the cursor for the page after it.
// PendingCount is what the Screener holds in total and is only answered by PendingPage.
type ClearancePage struct {
	Clearances   []generated.Clearance
	PendingCount int
	NextPage     string
}

// Pending answers the senders waiting to be screened, a page at a time.
//
// Each one carries the petitioner and the most recent entry they sent, so a caller can
// show who is asking and what they wrote without a second read. Pass the page token from
// a previous answer to walk the queue.
func (s *ClearancesService) Pending(ctx context.Context, page string) (summary *generated.ClearanceSummary, err error) {
	op := OperationInfo{
		Service: "Clearances", Operation: "GetClearances",
		ResourceType: "clearance", IsMutation: false,
	}

	err = s.client.instrument(ctx, op, func(ctx context.Context) error {
		resp, rerr := s.client.genClient().GetClearancesWithResponse(ctx, pendingClearancesParams(page))
		if rerr != nil {
			return rerr
		}
		if cerr := CheckResponse(resp.HTTPResponse); cerr != nil {
			return cerr
		}
		summary = resp.JSON200
		return nil
	})
	return summary, err
}

// PendingPage answers the same queue as Pending along with the cursor for the page after
// it, so a caller walking the queue is told when it has reached the end of it.
func (s *ClearancesService) PendingPage(ctx context.Context, page string) (result *ClearancePage, err error) {
	op := OperationInfo{
		Service: "Clearances", Operation: "GetClearances",
		ResourceType: "clearance", IsMutation: false,
	}

	err = s.client.instrument(ctx, op, func(ctx context.Context) error {
		resp, rerr := s.client.genClient().GetClearancesWithResponse(ctx, pendingClearancesParams(page))
		if rerr != nil {
			return rerr
		}
		if cerr := CheckResponse(resp.HTTPResponse); cerr != nil {
			return cerr
		}
		result = &ClearancePage{}
		if resp.JSON200 != nil {
			result.Clearances = resp.JSON200.Clearances
			result.PendingCount = int(resp.JSON200.PendingClearancesCount)
		}
		if resp.HTTPResponse != nil {
			result.NextPage = gearedPageFromLink(resp.HTTPResponse.Header.Get("Link"))
		}
		return nil
	})
	return result, err
}

// Screen answers the Screener for one sender. Status is ClearanceApproved or
// ClearanceDenied.
//
// The options are all optional: file everything they send into a box instead of the Imbox,
// mark what is already waiting as spam, or mark it seen so it does not arrive unread.
func (s *ClearancesService) Screen(ctx context.Context, clearanceID int64, status string, opts ScreenOptions) (clearance *generated.Clearance, err error) {
	op := OperationInfo{
		Service: "Clearances", Operation: "UpdateClearance",
		ResourceType: "clearance", IsMutation: true, ResourceID: clearanceID,
	}

	if err := validateClearanceStatus(status); err != nil {
		return nil, err
	}

	err = s.client.instrument(ctx, op, func(ctx context.Context) error {
		body := generated.UpdateClearanceJSONRequestBody{Status: status}
		if opts.DesignationBoxID != 0 {
			body.DesignationBoxId = opts.DesignationBoxID
		}
		if opts.Spam {
			spam := true
			body.Spam = &spam
		}
		if opts.MarkTopicsAsSeen {
			seen := true
			body.MarkTopicsAsSeen = &seen
		}

		resp, rerr := s.client.genClient().UpdateClearanceWithResponse(ctx, clearanceID, body)
		if rerr != nil {
			return rerr
		}
		if cerr := CheckResponse(resp.HTTPResponse); cerr != nil {
			return cerr
		}
		clearance = resp.JSON200
		return nil
	})
	return clearance, err
}

// ScreenOptions carries what to do beyond setting the status.
type ScreenOptions struct {
	// DesignationBoxID files everything the sender sends into that box rather than the Imbox.
	DesignationBoxID int64
	// Spam marks the topics already waiting as spam and trains the filter on them.
	Spam bool
	// MarkTopicsAsSeen screens the sender in without their waiting mail arriving unread.
	MarkTopicsAsSeen bool
}

// ScreenMany screens several senders at once and answers the clearances it changed.
//
// HEY answers 404 when none of the ids belong to the caller. A partial match succeeds and
// answers only what it touched, so compare the answer against what you sent.
func (s *ClearancesService) ScreenMany(ctx context.Context, clearanceIDs []int64, status string, spam bool) (clearances []generated.Clearance, err error) {
	op := OperationInfo{
		Service: "Clearances", Operation: "BulkUpdateClearances",
		ResourceType: "clearance", IsMutation: true,
	}

	if len(clearanceIDs) == 0 {
		return nil, ErrUsage("at least one clearance is required")
	}
	if err := validateClearanceStatus(status); err != nil {
		return nil, err
	}

	err = s.client.instrument(ctx, op, func(ctx context.Context) error {
		body := generated.BulkUpdateClearancesJSONRequestBody{
			Ids:    joinIDs(clearanceIDs),
			Status: status,
		}
		if spam {
			body.Spam = &spam
		}

		resp, rerr := s.client.genClient().BulkUpdateClearancesWithResponse(ctx, body)
		if rerr != nil {
			return rerr
		}
		if cerr := CheckResponse(resp.HTTPResponse); cerr != nil {
			return cerr
		}
		if resp.JSON200 != nil {
			clearances = resp.JSON200.Clearances
		}
		return nil
	})
	return clearances, err
}

// Punt clears the Screener. Everyone waiting is dropped and reexamined the next time they
// write, so nothing is decided for them.
//
// The work is queued, so the senders are still pending when this returns.
func (s *ClearancesService) Punt(ctx context.Context) error {
	op := OperationInfo{
		Service: "Clearances", Operation: "PuntClearances",
		ResourceType: "clearance", IsMutation: true,
	}

	return s.client.instrument(ctx, op, func(ctx context.Context) error {
		resp, err := s.client.genClient().PuntClearancesWithResponse(ctx)
		if err != nil {
			return err
		}
		return CheckResponse(resp.HTTPResponse)
	})
}

// Screened answers the senders already screened in or out, newest decision first, a page
// at a time.
func (s *ClearancesService) Screened(ctx context.Context, page string) (clearances []generated.Clearance, err error) {
	result, err := s.ScreenedPage(ctx, page)
	if err != nil || result == nil {
		return nil, err
	}
	return result.Clearances, nil
}

// ScreenedPage answers the same decisions as Screened along with the cursor for the page
// after it.
func (s *ClearancesService) ScreenedPage(ctx context.Context, page string) (result *ClearancePage, err error) {
	op := OperationInfo{
		Service: "Clearances", Operation: "GetMyClearances",
		ResourceType: "clearance", IsMutation: false,
	}

	err = s.client.instrument(ctx, op, func(ctx context.Context) error {
		params := &generated.GetMyClearancesParams{}
		if page != "" {
			params.Page = &page
		}

		resp, rerr := s.client.genClient().GetMyClearancesWithResponse(ctx, params)
		if rerr != nil {
			return rerr
		}
		if cerr := CheckResponse(resp.HTTPResponse); cerr != nil {
			return cerr
		}
		result = &ClearancePage{}
		if resp.JSON200 != nil {
			result.Clearances = resp.JSON200.Clearances
		}
		if resp.HTTPResponse != nil {
			result.NextPage = gearedPageFromLink(resp.HTTPResponse.Header.Get("Link"))
		}
		return nil
	})
	return result, err
}

// Rescreen changes its mind about a sender already screened in or out.
//
// This is the decided list, not the queue: Screen is what answers a pending sender.
func (s *ClearancesService) Rescreen(ctx context.Context, clearanceID int64, status string) (clearance *generated.Clearance, err error) {
	op := OperationInfo{
		Service: "Clearances", Operation: "UpdateMyClearance",
		ResourceType: "clearance", IsMutation: true, ResourceID: clearanceID,
	}

	if err := validateClearanceStatus(status); err != nil {
		return nil, err
	}

	err = s.client.instrument(ctx, op, func(ctx context.Context) error {
		body := generated.UpdateMyClearanceJSONRequestBody{Status: status}

		resp, rerr := s.client.genClient().UpdateMyClearanceWithResponse(ctx, clearanceID, body)
		if rerr != nil {
			return rerr
		}
		if cerr := CheckResponse(resp.HTTPResponse); cerr != nil {
			return cerr
		}
		clearance = resp.JSON200
		return nil
	})
	return clearance, err
}

func pendingClearancesParams(page string) *generated.GetClearancesParams {
	include := true
	params := &generated.GetClearancesParams{IncludeClearances: &include}
	if page != "" {
		params.Page = &page
	}
	return params
}

func validateClearanceStatus(status string) error {
	if status != ClearanceApproved && status != ClearanceDenied {
		return &Error{Code: CodeValidation, Message: fmt.Sprintf("clearance status must be %q or %q, got %q", ClearanceApproved, ClearanceDenied, status)}
	}
	return nil
}

File: rust/hey-sdk/src/services/clearances.rs
//! The Screener: who is waiting to be let in, and letting them in or turning them away.
//!
//! Clearing the Screener is the generated [`Clearances::punt`]. The work it starts is
//! queued, so everyone waiting is still pending when it answers; they are dropped and
//! reexamined the next time they write, so nothing is decided for them.

use std::str::FromStr;

use crate::error::{Error, ErrorCode};
use crate::generated::types::{
    BulkUpdateClearancesRequestContent, Clearance, ClearanceListResponse, ClearanceSummary,
    UpdateClearanceRequestContent, UpdateMyClearanceRequestContent,
};
use crate::pagination::Page;

pub use crate::generated::services::clearances::*;

/// The two decisions the Screener takes.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum ClearanceStatus {
    Approved,
    Denied,
}

/// What to do beyond setting the status. HEY reads each of these for truthiness, so one
/// left alone stays off the wire entirely rather than going out as a false.
#[derive(Debug, Clone, Default, PartialEq, Eq)]
pub struct ScreenOptions {
    /// Files everything the sender sends into that box rather than the Imbox.
    pub designation_box_id: Option<i64>,
    /// Marks the topics already waiting as spam and trains the filter on them.
    pub spam: bool,
    /// Screens the sender in without their waiting mail arriving unread.
    pub mark_topics_as_seen: bool,
}

impl ClearanceStatus {
    pub fn as_str(&self) -> &'static str {
        match self {
            ClearanceStatus::Approved => "approved",
            ClearanceStatus::Denied => "denied",
        }
    }
}

impl<'a> Clearances<'a> {
    /// How many senders are waiting, without fetching them.
    ///
    /// This is the cheap read HEY's own apps sync for the Screener badge. Use
    /// [`Clearances::pending`] for the senders themselves.
    pub async fn pending_count(&self) -> Result<i32, Error> {
        Ok(self
            .summary()
            .await?
            .pending_clearances_count
            .unwrap_or_default())
    }

    /// Everything HEY says about the Screener without the queue itself: how many senders
    /// are waiting, and the signed stream name to subscribe to on HEY's cable server to be
    /// told when that changes.
    pub async fn summary(&self) -> Result<ClearanceSummary, Error> {
        self.get(&GetClearancesParams::default())
            .await
            .map(Page::into_inner)
    }

    /// The senders waiting to be screened, a page at a time.
    ///
    /// Each one carries the petitioner and the most recent entry they sent, so a caller can
    /// show who is asking and what they wrote without a second read.
    pub async fn pending(&self, page: Option<&str>) -> Result<ClearanceSummary, Error> {
        self.pending_page(page).await.map(Page::into_inner)
    }

    /// The same queue as [`Clearances::pending`], keeping the cursor for the page after it
    /// so a caller walking the queue is told when it has reached the end.
    pub async fn pending_page(&self, page: Option<&str>) -> Result<Page<ClearanceSummary>, Error> {
        let params = GetClearancesParams {
            include_clearances: Some(true),
            page: page.map(str::to_string),
        };
        self.get(&params).await
    }

    /// Answers the Screener for one sender.
    pub async fn screen(
        &self,
        clearance_id: i64,
        status: ClearanceStatus,
        options: &ScreenOptions,
    ) -> Result<Clearance, Error> {
        let body = UpdateClearanceRequestContent {
            status: status.as_str().to_string(),
            designation_box_id: options.designation_box_id,
            spam: flag(options.spam),
            mark_topics_as_seen: flag(options.mark_topics_as_seen),
        };
        self.update(clearance_id, &body).await
    }

    /// Screens several senders at once and answers the clearances it changed.
    ///
    /// HEY answers 404 when none of the ids belong to the caller. A partial match succeeds
    /// and answers only what it touched, so compare the answer against what was sent.
    pub async fn screen_many(
        &self,
        clearance_ids: &[i64],
        status: ClearanceStatus,
        spam: bool,
    ) -> Result<Vec<Clearance>, Error> {
        if clearance_ids.is_empty() {
            return Err(Error::usage("at least one clearance is required"));
        }

        let body = BulkUpdateClearancesRequestContent {
            ids: join_ids(clearance_ids),
            status: status.as_str().to_string(),
            spam: flag(spam),
        };
        Ok(self
            .bulk_update(&body)
            .await?
            .clearances
            .unwrap_or_default())
    }

    /// The senders already screened in or out, newest decision first, a page at a time.
    pub async fn screened(&self, page: Option<&str>) -> Result<Vec<Clearance>, Error> {
        Ok(self
            .screened_page(page)
            .await?
            .into_inner()
            .clearances
            .unwrap_or_default())
    }

    /// The same decisions as [`Clearances::screened`], keeping the cursor for the page
    /// after it.
    pub async fn screened_page(
        &self,
        page: Option<&str>,
    ) -> Result<Page<ClearanceListResponse>, Error> {
        let params = GetMyClearancesParams {
            page: page.map(str::to_string),
        };
        self.get_my(&params).await
    }

    /// Changes its mind about a sender already screened in or out.
    ///
    /// This is the decided list, not the queue: [`Clearances::screen`] is what answers a
    /// pending sender.
    pub async fn rescreen(
        &self,
        clearance_id: i64,
        status: ClearanceStatus,
    ) -> Result<Clearance, Error> {
        let body = UpdateMyClearanceRequestContent {
            status: status.as_str().to_string(),
        };
        self.update_my(clearance_id, &body).await
    }
}

fn flag(value: bool) -> Option<bool> {
    if value { Some(true) } else { None }
}

fn join_ids(ids: &[i64]) -> String {
    ids.iter()
        .map(|id| id.to_string())
        .collect::<Vec<String>>()
        .join(",")
}

impl FromStr for ClearanceStatus {
    type Err = Error;

    fn from_str(source: &str) -> Result<ClearanceStatus, Error> {
        match source {
            "approved" => Ok(ClearanceStatus::Approved),
            "denied" => Ok(ClearanceStatus::Denied),
            _ => Err(Error::new(
                ErrorCode::Validation,
                format!("clearance status must be \"approved\" or \"denied\", got {source:?}"),
            )),
        }
    }
}
